bin/node/src/commands/start.rs

```diff
@@ -16,22 +18,41 @@ pub async fn start_node(config: NodeConfig) -> Result<()> {
 
     // Start store. The store endpoint is available after loading completes.
     let store = Store::init(store).await.context("Loading store")?;
-    join_set.spawn(async move { store.serve().await.context("Serving store") });
+    let store_id = join_set.spawn(async move { store.serve().await.context("Serving store") }).id();
 
     // Start block-producer. The block-producer's endpoint is available after loading completes.
     let block_producer =
         BlockProducer::init(block_producer).await.context("Loading block-producer")?;
-    join_set.spawn(async move { block_producer.serve().await.context("Serving block-producer") });
+    let block_producer_id = join_set
+        .spawn(async move { block_producer.serve().await.context("Serving block-producer") })
+        .id();
 
     // Start RPC component.
     let rpc = Rpc::init(rpc).await.context("Loading RPC")?;
-    join_set.spawn(async move { rpc.serve().await.context("Serving RPC") });
-
-    // block on all tasks
-    while let Some(res) = join_set.join_next().await {
-        // For now, if one of the components fails, crash the node
-        res??;
-    }
-
-    Ok(())
+    let rpc_id = join_set.spawn(async move { rpc.serve().await.context("Serving RPC") }).id();
+
+    // Lookup table so we can identify the failed component.
+    let component_ids = HashMap::from([
+        (store_id, "store"),
+        (block_producer_id, "block-producer"),
+        (rpc_id, "rpc"),
+    ]);
+
+    // SAFETY: The joinset is definitely not empty.
+    let component_result = join_set.join_next_with_id().await.unwrap();
+
+    // We expect components to run indefinitely, so we treat any return as fatal.
+    //
+    // Map all outcomes to an error, and provide component context.
+    let (id, err) = match component_result {
+        Ok((id, Ok(_))) => (id, Err(anyhow::anyhow!("Component completed unexpectedly"))),
+        Ok((id, Err(err))) => (id, Err(err)),
+        Err(join_err) => (join_err.id(), Err(join_err).context("Joining component task")),
+    };
+    let component = component_ids.get(&id).unwrap_or(&"unknown");
```

Comment:
It could be; but I thought this would result in a better error message than a panic in the case that we introduce a bug or forget to add a new component to the list.

With the panic we only get a `"unknown node component task ID"` message, whereas this way we get something more structured and critically includes the actual error:

```
Error: Component unknown failed

Caused by:
    0: <inner error>
```